stdlib/socket/from.go

```diff
@@ -0,0 +1,201 @@
+
+type FromSocketOpSpec struct {
+	URL     string `json:"url"`
+	Decoder string `json:"decoder"`
+}
+
+func init() {
+	fromSocketSignature := semantic.FunctionPolySignature{
+		Parameters: map[string]semantic.PolyType{
+			"url":     semantic.String,
+			"decoder": semantic.String,
+		},
+		Required: semantic.LabelSet{"url"},
+		Return:   flux.TableObjectType,
+	}
+
+	flux.RegisterPackageValue("socket", "from", flux.FunctionValue(FromSocketKind, createFromSocketOpSpec, fromSocketSignature))
+	flux.RegisterOpSpec(FromSocketKind, newFromSocketOp)
+	plan.RegisterProcedureSpec(FromSocketKind, newFromSocketProcedure, FromSocketKind)
+	execute.RegisterSource(FromSocketKind, createFromSocketSource)
+}
+
+type administrationTimeProvider struct {
+	a execute.Administration
+}
+
+func (a *administrationTimeProvider) GetCurrentTime() values.Time {
+	return a.a.ResolveTime(flux.Time{
+		IsRelative: false,
+		Absolute:   time.Now(),
+	})
+}
+
+var decoders = map[string]bool{
+	"json": false, // TODO(affo) should flatten json into table
+	"csv":  true,
+	"line": true,
+}
+
+func createFromSocketOpSpec(args flux.Arguments, a *flux.Administration) (flux.OperationSpec, error) {
+	spec := new(FromSocketOpSpec)
+
+	if url, err := args.GetRequiredString("url"); err != nil {
+		return nil, err
+	} else {
+		spec.URL = url
+	}
+
+	if d, ok, err := args.GetString("decoder"); err != nil {
+		return nil, err
+	} else if ok {
+		spec.Decoder = d
+	} else {
+		spec.Decoder = "csv"
+	}
+
+	if !decoders[spec.Decoder] {
+		return nil, fmt.Errorf("unknow decoder %v: decoder must be one of %v", spec.Decoder, decoders)
+	}
+
+	return spec, nil
+}
+
+func newFromSocketOp() flux.OperationSpec {
+	return new(FromSocketOpSpec)
+}
+
+func (s *FromSocketOpSpec) Kind() flux.OperationKind {
+	return FromSocketKind
+}
+
+type FromSocketProcedureSpec struct {
+	plan.DefaultCost
+	URL     string
+	Decoder string
+}
+
+func newFromSocketProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*FromSocketOpSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+
+	return &FromSocketProcedureSpec{
+		URL:     spec.URL,
+		Decoder: spec.Decoder,
+	}, nil
+}
+
+func (s *FromSocketProcedureSpec) Kind() plan.ProcedureKind {
+	return FromSocketKind
+}
+
+func (s *FromSocketProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(FromSocketProcedureSpec)
+	ns.URL = s.URL
+	ns.Decoder = s.Decoder
+	return ns
+}
+
+func createFromSocketSource(s plan.ProcedureSpec, dsid execute.DatasetID, a execute.Administration) (execute.Source, error) {
+	spec, ok := s.(*FromSocketProcedureSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", s)
+	}
+
+	url := spec.URL
+	network := "tcp"
+	i := strings.LastIndex(url, "://")
+	// if :// is at the beginning or end,
+	// we pass the url as it is, and use "tcp" as default.
+	if i > 0 && i < len(url)-3 {
+		network = url[:i]
+		url = url[i+3:]
+	}
+	conn, err := net.Dial(network, url)
```

Comment:
What about `file://`? The net.Dial function will not open the file for you.

Reply:
Mmmm, that's ok, but this is a SocketSource, should I account for it or let net.Dial return an error?

Reply:
Lets only pass networks into net.Dial that we know will work, that way we can return our own more meaningful error messages.

Reply:
Agreed. In the future, the source could serve as a base class for every source that needs a `io.ReadCloser`. And so being used by:
 - `csv.from`
 - `file.from`
 - `stdio.from`
 - `net.from`

`csv` could also become `file.from(decoder:"csv", ...)`.